### ISSUE TITLE
Incorrect `ThetaJoin` combiner function

### ISSUE THREAD

**Issue by wemrysi** [quasar-analytics/quasar#1515]
Following is the `LogicalPlan` and a problematic `ThetaJoin` from the `QScript` for the query `select sum(pop) from /explain/data` (The full `QScript` is quite large and so I have elided all but the bit of the problematic `ThetaJoin`).

```
[info] Optimized
[info] Invoke(SQUASH)
[info] ╰─ Invoke(MAKE_OBJECT)
[info]    ├─ Str("0")
[info]    ╰─ Invoke(SUM)
[info]       ╰─ ObjectProject(/explain/data["pop"])

[info] Typechecked
[info] Invoke(SQUASH)
[info] ╰─ Invoke(MAKE_OBJECT)
[info]    ├─ Str("0")
[info]    ╰─ Invoke(SUM)
[info]       ╰─ Let('__check0)
[info]          ├─ Read(/explain/data)
[info]          ╰─ Typecheck(Type(Obj(Map(),Some(Top))))
[info]             ├─ Free('__check0)
[info]             ├─ Let('__check1)
[info]             │  ├─ ObjectProject('__check0["pop"])
[info]             │  ╰─ Typecheck(Type(Coproduct(Coproduct(Int,Dec),Interval)))
[info]             │     ├─ Free('__check1)
[info]             │     ├─ Free('__check1)
[info]             │     ╰─ Constant(NA)
[info]             ╰─ Constant(NA)

ThetaJoin(
    Unreferenced,
    Map(
        ShiftedRead(ShiftedRead(/quasar-test/run_64541428/explain/data, IdStatus(IncludeId))),
        ProjectField(ProjectIndex(SrcHole, Constant(Int(1))), Constant(Str(pop)))),
    Map(
        ShiftedRead(ShiftedRead(/quasar-test/run_64541428/explain/data, IdStatus(IncludeId))),
        ProjectField(ProjectIndex(SrcHole, Constant(Int(1))), Constant(Str(pop)))),
    Constant(Bool(true)),
    Inner,
    ConcatArrays(
        MakeArray(
            ConcatArrays(
                ConcatArrays(
                    ConcatArrays(
                        ConcatArrays(
                            MakeArray(
                                MakeMap(
                                    Constant(Str(j)), 
                                    ConcatArrays(
                                        MakeArray(
                                            MakeMap(
                                                Constant(Str(j)), 
                                                ConcatArrays(
                                                    MakeArray(
                                                        MakeMap(
                                                            Constant(Str(j)), 
                                                            ConcatArrays(
                                                                MakeArray(
                                                                    MakeMap(
                                                                        Constant(Str(m)), 
                                                                        ProjectIndex(ProjectIndex(ProjectIndex(LeftSide, Constant(Int(0))), Constant(Int(1))), Constant(Int(0)))))
```
The issue is in the `ThetaJoin` combiner expression, you can see the left branch consists of the `DataSet` of `pop` fields projected out of the rhs of the `ShiftedRead` and then in the combiner, it attempts to index into that value (`LeftSide`) as if it were an array.

**Comment by sellout** [comment]
It looks like you elided part of that `ConcatArrays`? If I’m not mistaken, the second argument to it should have the condition in it. In any case, I’ve gotten it fixed locally, but still have a couple others things to improve with joins before I’m happy with submitting the PR.

I guess I should be happy that all the connectors are actually able to test joins at this point 😆

**Comment by wemrysi** [comment]
@sellout Yeah, I probably did in the spirit of minimizing the problem, sorry.

This particular query doesn't even involve a `ThetaJoin` any longer, so that's nice ;)

Let me know if it would be helpful to have the full `QScript`, I can checkout a branch prior to your recent commits and rerun the query.

**Comment by wemrysi** [comment]
Whoops, didn't mean to close.

**Comment by sellout** [comment]
So, giving a quick glance. I think the query in this ticket,
```sql
select sum(pop) from `/explain/data`
```
now behaves properly, right?

I’ll keep it open for the general join issue, but I’m fine on noisy ThetaJoin QScript examples for now 😉

**Comment by wemrysi** [comment]
Oh, yes, sorry that's what I meant by "doesn't even involve a `ThetaJoin`", should have been more clear.

**Comment by sellout** [comment]
I figured you might, but there’s a wide gulf between those two statements, so I wanted to check 😄

**Comment by sellout** [comment]
This was fixed in 12.4.3.
